### PR TITLE
Fix array length in symex constant string propagation [TG-9048]

### PR DESCRIPTION
**Pull request by romainbrenguier**
This adds an  assumption on the length in assign_string_constant before we do the actual assignment. This is to make sure that the length of the array before the assignment
isn't unconstrained. Otherwise this array could be set to be arbitrarily large
by the solver which would cause invalid traces.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-symex/goto_symex.h [hunk 1]
```diff
@@ -590,6 +590,9 @@ class goto_symext
   /// "abc", the symbol will be named "abc_constant_char_array"). Then, the
   /// expression `&sym[0]` is assigned to `char_array` (assuming `sym` denotes
   /// the symbol holding the string data given by `new_char_array`.
+  /// The assignment is preceeded by a assume statements ensuring `length`
```
Comment (owen-mc-diffblue) [inline src/goto-symex/goto_symex.h]: "a assume statements" -> "an assume statement"

File: src/goto-symex/goto_symex.h [hunk 2]
```diff
@@ -590,6 +590,9 @@ class goto_symext
...
+  /// before this call was zero, this is to avoid letting the previous array
```
Comment (owen-mc-diffblue) [inline src/goto-symex/goto_symex.h]: "letting the previous array unconstrained" -> "letting the previous array be unconstrained" or "leaving the previous array unconstrained"

File: jbmc/regression/jbmc-strings/StringBuilderConstructors01/test-no-arg-fail.desc [hunk 3]
```diff
@@ -1,8 +1,12 @@
...
+--function StringBuilderConstructors01.noArgFail --max-nondet-string-length 1000 --trace
 ^EXIT=10$
 ^SIGNAL=0$
 ^VERIFICATION FAILED$
+dynamic_object2=\{\s*\}
 --
 ^warning: ignoring
+--
+The check for dynamic_object2 is to make sure the array is created empty and
+is not given arbitrary content before its final assignment.
```
Comment (owen-mc-diffblue) [inline jbmc/regression/jbmc-strings/StringBuilderConstructors01/test-no-arg-fail.desc]: missing newline

**Review by owen-mc-diffblue — APPROVED** [review]

**Review by smowton — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: e03ca3d085e0de2ece371401a435c11c98d2703b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/123901901
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 334714a4c1437053c0b0e84cca7c231956645a71).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/123906978

**Pull request merged**